### ISSUE TITLE
How to use moveToFailed and moveToCompleted

### ISSUE THREAD

**Issue by grapevineai** [taskforcesh/bullmq#327]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%2440%20Funded-%2300A156.svg" />](https://issuehunt.io/r/taskforcesh/bullmq/issues/327)

These two functions are different from v3 and I just don't know what to set for the token.

`job.moveToFailed(new Error('failed message'), ???);`

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [<img src='https://avatars3.githubusercontent.com/u/28606840?v=4' alt='grapevineai' width=24 height=24> grapevineai](https://issuehunt.io/u/grapevineai) ($40.00)

### Submitted pull Requests
- [#354 feat: add support for manually processing jobs fixes #327](https://issuehunt.io/r/taskforcesh/bullmq/pull/354)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/taskforcesh/bullmq/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by NDT36** [comment]
![image](https://user-images.githubusercontent.com/44062844/101620099-b7c20680-3a46-11eb-8d9c-4b4744c85db9.png)
here

**Comment by grapevineai** [comment]
@NDT36 thanks but not related... please read carefully the issue!

**Comment by DaniSchenk** [comment]
@grapevineai the `Queue` instance your `Worker` is working on has a `token` property. Providing it in `moveToFailed()` seems to work in my case.

**Comment by grapevineai** [comment]
@DaniSchenk Got the same issue...

@manast Any recommendation here?

**Comment by manast** [comment]
@grapevineai these methods are considered private in BullMQ, however we could make them public as a new feature, so that we have proper tests and can guarantee that they work.

**Comment by manast** [comment]
what is your use case by the way?

**Comment by DaniSchenk** [comment]
> what is your use case by the way?

I start a long running background task via HTTP request using `job1`. The immediate response would be an ID. I'm using this ID in `job2`, `job3`, `job...` to check the status/process of the request made in `job1`.

Let's say `job1` fails. One of the status checking jobs will observe this failure. In case of a failure I want to move `job1` to the list of failed jobs.
It would make it easier to track all failed jobs in e.g. a dashboard. All failed jobs would be in one list and no further job inspections are needed to find additional failed jobs.

**Comment by manast** [comment]
@DaniSchenk you say that job1 fails but it was moved to the completed list anyway because its creation was successful. Why can't you just throw an exception if the job1 fails instead of completing before you know if it has failed or not?

**Comment by grapevineai** [comment]
@manast My use case is the following:
- I have delayed jobs pending
- I realize my jobs won't be able to run after few of them completed
- I want to "discard" the remaining jobs and move them to failed (so I can keep track)

**Comment by DaniSchenk** [comment]
> @DaniSchenk you say that job1 fails but it was moved to the completed list anyway because its creation was successful. Why can't you just throw an exception if the job1 fails instead of completing before you know if it has failed or not?

@manast I think I explained my use case a bit ambiguous. I'll try it another example:

`job1` makes HTTP request to another service
Request was successful -> `job1` is moved to completed

**five minutes later:**
`job2` makes HTTP request to the other service to check status of a process initiated by `job1`
Request was successful -> `job2` is moved to completed

**But** response data of `job2` indicates a failure of a process initiated by `job1`. So the request `job1` made is responsible for the failure. This is why i want to move `job1` to failed.

**Comment by grapevineai** [comment]
@manast Any chance we can get that public? It's public in v3 why was it removed?

**Comment by manast** [comment]
It has not been removed, BullMQ is written from scratch and this functionality has not been ported/implemented yet. It is fairly straightforward to implement, yet some tests and documentation is also needed. I create a bounty for this:
https://issuehunt.io/r/taskforcesh/bullmq/issues/327

**Comment by grapevineai** [comment]
@manast I'm willing to contribute to this bounty, but I need to make sure I can achieve the ultimate result with that. It seems that `getNextJob()` is also not part of the new API. How can we use BullMQ if the job is performed by a third party?

Let's say, my clients connect to the server to get the next job. They perform the job on their own and send the result once done. I thought that by using getNextJob, moveToFailed and moveToCompleted I could achieve that. A similar use case is exposed on v3: https://github.com/OptimalBits/bull/blob/develop/PATTERNS.md#manually-fetching-jobs

How can this be achieved in v4? If I have the guarantee I can perform that, I'd be happy to contribute!

**Comment by manast** [comment]
Yes, they will need to call ```getNextJob()``` otherwise there is no job to move since moving jobs that are already being automatically being processed by workers is not a good idea. But of course making ```getNextJob``` public will also be part of this implementation.

**Comment by grapevineai** [comment]
@manast How long will it take to get these three functions done (from when the bounty is fully covered)? Is it a matter of days/weeks/months?

**Comment by manast** [comment]
days

**Comment by manast** [comment]
Documentation: https://docs.bullmq.io/patterns/manually-fetching-jobs

**Comment by DaniSchenk** [comment]
Thank you both!

One will have to manually process the queue to use `moveToFailed()`?

My approach was to use it like this (simplified) in process functions:
```
worker.on("completed", async (job: job) => {
  [ ... ]
  // move some job in job.data to failed
  await moveJobToFailedById(job.data.someOtherJobId);
});

async moveJobToFailedById(jobId: string) {
  const job = await Job.fromId(myQueue, jobId);
  await job.moveToFailed(new Error(errorMsg), myQueue.token);
}
```

**Comment by manast** [comment]
You cannot fail a job that has already completed.

**Comment by DaniSchenk** [comment]
Oh, ok. Thanks for clearing that up.

**Comment by grapevineai** [comment]
@manast Thanks, works like a charm!

One last question, I'd like to extend the `lockDuration` on request. Since it's not possible to update that on the fly, I thought I could be able to remove the job, add the same job and finally attribute it again with an extended `lockDuration`. The problem is that I can't pick the newly created job from the queue (the only way to manually get the job is to call `getNextJob()`).

Is there a way to solve this?

**Comment by manast** [comment]
I released 1.14.0 with a new job method ```extendLock``` that you can use to extend the lock: https://docs.bullmq.io/patterns/manually-fetching-jobs

**Comment by grapevineai** [comment]
That's awesome, thank you so much!

**Comment by grapevineai** [comment]
That's awesome, thank you so much!

**Comment by JaroslavPrt** [comment]
hi @manast, is it works ok with concurrency?

**Comment by manast** [comment]
@JaroslavPrt if you fetch the jobs manually then you have to implement concurrency manually as well.

**Comment by grapevineai** [comment]
@manast Somehow, the lockduration doesn't work properly... First of all, the job stays in active state for 1 minute no matter what lockduration is set. Second, if the lockduration is too low or too high (I tries with 10sec and 10min), the job moves to the fail state after 1 minute.
I don't know exactly what's going on here but the lockduration doesn't seem to work at all.

v1.14.0

**Comment by manast** [comment]
@grapevineai sorry for the late response. I will investigate this issue.

**Comment by manast** [comment]
@grapevineai ok, I think I understand what is going on.

The ```QueueScheduler``` class is responsible for moving stalled jobs from active to wait if it considers the job stalled. However the mechanism for moving back to wait is independent of the locking. It will check once every ```stalledInterval```  which by default is 30 seconds. The first time it will put all the active jobs in a special set as potential stalled jobs. The next time it will check in that set again, if any job is still there then it will consider it stalled and move it back to wait. This mechanism is not optimal when manually moving jobs as you do, and I need to think how to improve it. However you should be able to easily workaround it: when you instantiate ```QueueScheduler``` pass a ```stalledInterval``` option with a value half of your lock duration. By default it is 1 minute thats why default ```stalledinterval``` is 30 seconds...

I hope this helps.

**Comment by grapevineai** [comment]
Hey, thanks for giving me a workaround. This is unfortunately not satisfying in my case since the `extendLock` won't work (if I understand correctly...) :(

**Comment by manast** [comment]
@grapevineai yes, extendLock will work, why can't you use the workaround I suggested?

**Comment by grapevineai** [comment]
@manast Then I think I didn't understand the problem... I thought that the stalled job function would make the `extendLock` useless since I have to set a specific time as `stalledInterval`. Let me try the workaround then! Thanks!

**Comment by grapevineai** [comment]
@manast I have tested more extensively your workaround. Unfortunately it doesn't fully work. It goes back to waiting the first time it timed-out but the second time it goes to fails. I have tried to set `attempts: Number.MAX_SAFE_INTEGER` when I initially add the task but it didn't help. In addition, the behavior is not consistent and sometimes the task stays in active state after the timeout ended.

**Comment by manast** [comment]
@grapevineai you need to specify the ```maxStalledCount``` option in the QueueScheduler which is 1 by default. However, you should not get any stalled jobs if you are calling ```extendLock``` before the locks expires.

**Comment by grapevineai** [comment]
@manast Thanks for the precision! There is another issue which is a real bug this time...I have added a couple of tasks in my queue. I let my first task expire and automatically get the next task. I let the second task expire and get the first task back. I try to complete the first task received for the second time and get the following error (the app crashes):

```
ReplyError: ERR Error running script (call to f_c09b01b58cdf1bb980c4b01fe186dd5e236a6770): @user_script:94: ERR value is not an integer or out of range
    at parseError (/app/node_modules/redis-parser/lib/parser.js:179:12)
    at parseType (/app/node_modules/redis-parser/lib/parser.js:302:14) {
  command: {
    name: 'evalsha',
    args: [
      'c09b01b58cdf1bb980c4b01fe186dd5e236a6770',
      '7',
      'annot:bull:clear:active',
      'annot:bull:clear:completed',
      'annot:bull:clear:972b62ab-d690-4c48-b981-611f39fa1091',
      'annot:bull:clear:wait',
      'annot:bull:clear:priority',
      'annot:bull:clear:events',
      'annot:bull:clear:meta',
      '972b62ab-d690-4c48-b981-611f39fa1091',
      '1614089835495',
      'returnvalue',
      '"completed by: test@gmail.com at: 1614089835494"',
      'completed',
      '',
      '{"jobId":"972b62ab-d690-4c48-b981-611f39fa1091","val":"\\"completed by: test@gmail.com at: 1614089835494\\""}',
      '1',
      'annot:bull:clear:',
      'test@gmail.com',
      ''
    ]
  }
}

Process finished with exit code 1
```

**Comment by manast** [comment]
@grapevineai is it possible that you can post a code snippet that reproduces that exact error please?

**Comment by grapevineai** [comment]
Sure! I initialize my queue:

```javascript
const connection = {prefix:'annot:bull', connection: {port: address.port, host: address.hostname, password: process.env.REDIS_PASS}};
const queue = new Queue('myqueue', connection).on("error", (err) =>{console.log(err);});
new QueueScheduler('myqueue', Object.assign({stalledInterval: 60000, maxStalledCount:Number.MAX_SAFE_INTEGER}, connection)).on("error", (err) =>{console.log(err);});
```

I add two tasks as follows:

```javascript
await queue.add('req:' + requested._id, {request: requested._id, thread: thread._id, new:true}, {
    jobId: requested._id,
    priority: requested.priority,
    attempts: Number.MAX_SAFE_INTEGER
}).catch(err => {
    console.log(err);
});
```

and get the next task:

```javascript
const worker = new Worker('myqueue', null, Object.assign({lockDuration: 120000}, connection));
const job = await worker.getNextJob('test@gmail.com');
```

I let it expire, get the next one right away, let the second one expire and receive the first one again. For the second time the first task was received, I complete it:

```javascript
const job = await clearQueue.getJob(requested._id);
if(job){
    await job.moveToCompleted('completed by: ' + 'test@gmail.com' + ' at: ' + Date.now(), 'test@gmail.com');
}
```

and get the error mentioned above.

**Comment by grapevineai** [comment]
@manast Sorry I forgot to add your name in the previous comment

**Comment by grapevineai** [comment]
@manast Any chance this can be investigated ASAP?

**Comment by manast** [comment]
sorry, for late response. Basically I need a piece of code that reproduces the issue that I can just run directly as is. The current code makes many assumptions and If I try to replicate them besides taking a lot of my time it may also result in a different case. So please create a standalone .js file that I can just run with ```node test.js``` and that reproduces the issue and I can look into it right away.

**Comment by grapevineai** [comment]
@manast Of course, sorry about that...

```javascript
const { Queue, QueueScheduler, Worker } = require('bullmq');

const connection = {connection: {port: '6379', host: 'localhost'}};
const queue = new Queue('myqueue', connection).on("error", (err) =>{console.log(err);});
new QueueScheduler('myqueue', Object.assign({stalledInterval: 1000, maxStalledCount:Number.MAX_SAFE_INTEGER}, connection)).on("error", (err) =>{console.log(err);});

(async () =>{
    await queue.add('test1', {}, {
        attempts: Number.MAX_SAFE_INTEGER
    }).catch(err => {
        console.log(err);
    });
    await queue.add('test2', {}, {
        attempts: Number.MAX_SAFE_INTEGER
    }).catch(err => {
        console.log(err);
    });
    await queue.add('test3', {}, {
        attempts: Number.MAX_SAFE_INTEGER
    }).catch(err => {
        console.log(err);
    });

    async function getJob(complete){
        const worker = new Worker('myqueue', null, Object.assign({lockDuration: 2000}, connection));
        const next = await worker.getNextJob('test@gmail.com');

        if(complete){
            const job = await queue.getJob(next.id);
            if(job){
                console.log('complete: ' + next.id);
                await job.moveToCompleted('completed by: ' + 'test@gmail.com' + ' at: ' + Date.now(), 'test@gmail.com');
            }
        }
    }


    getJob(false);
    setTimeout(getJob, 1000, false);
    setTimeout(getJob, 2000, true);
})()
```

**Comment by manast** [comment]
I could reproduce the issue. The problem is that the Job instance used for the call ```moveToCompleted``` has been created in the queue instance that does not have any value for the ```lockDuration``` option. If you use ```next``` instead it works. In general thats the job instance you want to use anyway. I have to figure out how to change the current implementation so that such problem can never happen... probably have a different Job instances for workers and for queues. 

```ts
   async function getJob(complete){
        const worker = new Worker('myqueue', null, Object.assign({lockDuration: 2000}, connection));
        const next = await worker.getNextJob('test@gmail.com');

        if(complete && next){
                console.log('complete: ' + next.id);
                await next.moveToCompleted('completed by: ' + 'test@gmail.com' + ' at: ' + Date.now(), 'test@gmail.com');
            }
        }
    }
```

**Comment by manast** [comment]
In that case we need a new getJob method in the Worker class.

**Comment by manast** [comment]
you can try this as a workaround:
```ts
        const worker = new Worker('myqueue', null, Object.assign({lockDuration: 2000}, connection));
        const job = Job.fromId(worker, jobId);
```

**Comment by grapevineai** [comment]
@manast Thanks for the workaround! The code is more complex and quite nasty but it works... any chance this will be cleaned up soon?

**Comment by manast** [comment]
why is it nasty? 
The only improvement that can be done here is to expose getJob in the Worker class, so it would look like this:
```ts
 const worker = new Worker('myqueue', null, Object.assign({lockDuration: 2000}, connection));
 const job = worker.getJob(jobId);
```

In general calling getJob should not be used for manually moving the jobs, instead ```getNextJob```  is the way to do it, but since it is not possible in your case this is the only workaround I can come up with.

**Comment by grapevineai** [comment]
In my case it's nasty because the `lockDuration` has to be set for `getNextJob` and `getJob`. What happens if the numbers are different by mistake? There might be some strange behavior...

**Comment by manast** [comment]
Just use a ENV variable that is configured via a centralized parameter store.

**Comment by IoannMelnychuk** [comment]
> why is it nasty? The only improvement that can be done here is to expose getJob in the Worker class, so it would look like this:
> 
> ```ts
>  const worker = new Worker('myqueue', null, Object.assign({lockDuration: 2000}, connection));
>  const job = worker.getJob(jobId);
> ```
> 
> In general calling `getJob` should not be used for manually moving the jobs, instead `getNextJob` is the way to do it, but since it is not possible in your case this is the only workaround I can come up with.

Is it still actual info?
I've tried to use `fromId` but got `Missing lock for the job XXXX`. Is it possible to move job by id to completed manually?

**Comment by manast** [comment]
https://docs.bullmq.io/patterns/manually-fetching-jobs

**Comment by IoannMelnychuk** [comment]
@manast yep, I've seen that part of docs. But there is no info about getting job by id for manual processing. In my case, when I use `fromId` with passing Worker instance, it throw an error `Missing lock for the job XXXX`. Is there a way to do something like - `fromId(jobId, worker, token)` or `worker.getJob(jobId, token)`?

**Comment by manast** [comment]
@IoannMelnychuk the docs explains how to work with manual jobs. Not sure what you are trying to do though... but yes, you are getting the error because you cannot just ```moveToCompleted``` for jobs that you do not own the lock for.

**Comment by IoannMelnychuk** [comment]
@manast how to own the lock for the job?
Not with `getNextJob` method, but with more accurate way (fromId, getJob, etc)?

**Comment by rtroncoso** [comment]
Bump, I have a similar use case as @IoannMelnychuk where my job sends HTTP requests to an external service and I need to update the status of the job from this external server once it finishes the process. Steps:

- Send HTTP request from job processor, wait using `job.waitUntilFinished()`
- Handle HTTP request from external service, perform large process
- Once the large process finishes, obtain the job using `Job.fromId()` and afterwards update the status of the job using `job.moveToCompleted()` or `job.moveToFailed()` accordingly

Result:
```Missing lock for job 13. finished```

Is there a way of doing this without manually processing jobs? Thanks!

**Comment by manast** [comment]
@rtroncoso I wonder if your case could not be implemented easier without "waitUntilFinished()". That method is very seldom  needed and you can do it better in a different way. For example instead of waiting, handle the HTTP request from the external service in a different queue, so you just finished the processing of the first queue by adding the next job. Then you have access to everything and locks, etc are handled automatically by BullMQ.

**Comment by manast** [comment]
Btw, there are several articles that can give you hints on how to construct it in a better way here: https://blog.taskforce.sh/

**Comment by rtroncoso** [comment]
Thanks for the reply @manast ! That's definitely the way I'd go for it if I was doing this from scratch, however, we are currently migrating from a previous scheduler system which used HTTP calls to our main core-api service (which also depend on different modules and internal controllers that are already set-up and working).

I tried fiddling around with manually processing jobs and sending the token through the HTTP request itself. I was able to correctly manage the status of the job but, as I expected, there is no way to signal the `queueEvents` interface that the process is completed from an external service.

For now we are going to keep these external jobs out of the migration, but it would be great if there was a way to manage these kind of implementations (given the amount of people with similar use cases as well) and I'd be willing to help in anything needed for this.
